### PR TITLE
Add update service page for both claims and placements

### PR DESCRIPTION
**Pull request by gms-gs**
## Context

We are making this change to allow us to update users for each service with news

## Changes proposed in this pull request

Add MVC for this

## Guidance to review

Run tests and or run app loccally

## Link to Trello card

https://trello.com/c/1qgE9sRc/44-news-and-updates-list-page

## Things to check

- [ ] If the code removes any existing feature flags, a data migration has also been added to delete the entry from the database
- [x] This code does not rely on migrations in the same Pull Request
- [ ] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [ ] If this code adds a column to the DB, decide whether it needs to be in analytics yml file or analytics blocklist
- [ ] API release notes have been updated if necessary
- [ ] If it adds a significant user-facing change, is it documented in the [CHANGELOG](CHANGELOG.md)?
- [ ] Required environment variables have been updated or [added to the Azure KeyVault](/docs/environment-variables.md#deploy-pipeline)

## Screenshots
Screenshot 2023-12-14 at 15 02 11
Screenshot 2023-12-14 at 15 18 44

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -100,3 +100,5 @@ group :test, :development do
   gem "rspec"
   gem "rspec-rails"
 end
+
+gem "redcarpet", "~> 3.6"
```
Comment (JamieCleare2525) [inline Gemfile:104]: Not a blocker, but could you move this higher up the Gemfile?
Just so there is a grouping of "Gems used in all environments", "Gems used in Development", "Gems used in Test", etc.

**Review by JamieCleare2525 — APPROVED** [review]
Ran locally and everything working as expected.

File: db/placements_service_updates.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# Example structure for adding service updates
+
+# - date: '2023-11-29'
```
Comment (SocFoot) [inline db/placements_service_updates.yml:3]: same

File: db/claims_service_updates.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Example structure for adding service updates
+
+# - date: '2023-11-29'
```
Comment (SocFoot) [inline db/claims_service_updates.yml:3]: Can we have a structure of what the yaml look with type rather than example.

File: spec/models/service_update_spec.rb [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+        updates = ServiceUpdate.where(service: :placements)
+        expect(updates).to eq({})
+      end
+    end
+  end
+
+  describe ".yaml_file" do
+    it "returns claims YAML file path" do
+      file_path = ServiceUpdate.yaml_file(service: :claims)
+      expect(file_path).to eq(Rails.root.join("db/claims_service_updates.yml"))
```
Comment (SocFoot) [inline spec/models/service_update_spec.rb:57]: You probably want to read from a fixture rather than a file that actually do something

File: spec/models/service_update_spec.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    end
+
+    it "returns placements YAML file path" do
+      file_path = ServiceUpdate.yaml_file(service: :placements)
+      expect(file_path).to eq(
+        Rails.root.join("db/placements_service_updates.yml")
```
Comment (SocFoot) [inline spec/models/service_update_spec.rb:63]: Same

File: app/assets/components/service_update/view.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+class ServiceUpdate::View < GovukComponent::Base
```
Comment (CatalinVoineag) [inline app/assets/components/service_update/view.rb:3]: I'm not that familiar with these Gov components but why do we put them in the assets folder? Why can't we follow the `view component` pattern and have them in `app/components`?

File: config/routes.rb [hunk 7]
```diff
@@ -14,6 +14,11 @@
   get("/auth/developer/sign-out", to: "sessions#signout")
   post("/auth/developer/callback", to: "sessions#callback")
 
+  resources :service_updates,
+            only: %i[index],
+            path: "service-updates",
+            controller: "service_updates"
```
Comment (CatalinVoineag) [inline config/routes.rb:20]: Not sure if the controller param is necessary if the resource is named the same as the controller 🤷🏻

File: app/assets/components/service_update/view.rb [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+
+  TITLE_TAGS = %w[h1 h2 h3 h4 h5 h6].freeze
+  TITLE_CLASS = "govuk-heading-m govuk-!-margin-bottom-2"
+
+  def initialize(service_update:, title_tag: "h2")
+    @service_update = service_update
+    @title_tag = title_tag
+  end
+
+  def render?
```
Comment (SocFoot) [inline app/assets/components/service_update/view.rb:16]: Seems that you want to use `alias`

File: app/assets/components/service_update/view.rb [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+  end
+
+  def date_pretty
+    date.to_date.strftime("%e %B %Y")
+  end
+
+  def content_html
+    custom_render =
+      Redcarpet::Render::HTML.new(link_attributes: { class: "govuk-link" })
+    Redcarpet::Markdown.new(custom_render).render(content).html_safe
```
Comment (SocFoot) [inline app/assets/components/service_update/view.rb:35]: you can pass `custom_render` directly

File: app/models/service_update.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+
+    return {} unless updates
+
+    updates.map { |params| new(params) }
+  end
+
+  def self.yaml_file(service:)
+    case service
+    when :claims
+      Rails.root.join(CLAIMS_SERVICE_UPDATES_YAML_FILE)
```
Comment (SocFoot) [inline app/models/service_update.rb:29]: Is the `Rails.root.join` useful here? Seems like `CLAIMS_SERVICE_UPDATES_YAML_FILE` already used it

File: app/models/service_update.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+
+  include ActiveModel::Model
+
+  attr_accessor :date, :title, :content
+
+  def id
+    title.parameterize
+  end
+
+  def self.where(service:)
```
Comment (SocFoot) [inline app/models/service_update.rb:17]: In rails where is used for return value that would be a list. I think what you want is a get/find

File: app/models/service_update.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class ServiceUpdate
+  CLAIMS_SERVICE_UPDATES_YAML_FILE =
+    Rails.root.join("db/claims_service_updates.yml").freeze
+  PLACEMENTS_SERVICE_UPDATES_YAML_FILE =
+    Rails.root.join("db/placements_service_updates.yml").freeze
+
+  include ActiveModel::Model
```
Comment (SocFoot) [inline app/models/service_update.rb:9]: Do you need this? Doesn't seems that you are using active model.

File: app/models/service_update.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+  attr_accessor :date, :title, :content
```
Comment (SocFoot) [inline app/models/service_update.rb:11]: But if you want to use it you probably want to remove attr_accessor.
```suggestion
class ServiceUpdate
  include ActiveModel::Attributes

  attribute :date, :string
  ...
end
```

File: app/models/service_update.rb [hunk 14]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+class ServiceUpdate
```
Comment (SocFoot) [inline app/models/service_update.rb:3]: The name update will probably be confusing down the line. Can you use a more specific name like user feed(probably ask product people)

**Review by SocFoot — CHANGES_REQUESTED** [review]
Good work a lot of files a lot of comments

File: app/assets/components/service_update/view.rb [hunk 15]
```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
```
Comment (elceebee) [inline app/assets/components/service_update/view.rb:1]: I don't think we need this -- it was discussed in an earlier PR.

File: app/models/service_update.rb [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+  def id
+    title.parameterize
+  end
+
+  def self.where(service:)
+    path = yaml_file(service:)
+    updates = YAML.load_file(path) if path
+
+    return {} unless updates
```
Comment (Nitemaeric) [inline app/models/service_update.rb:21]:
```suggestion
    return [] unless updates
```

File: app/models/service_update.rb [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+    updates.map { |params| new(params) }
```
Comment (Nitemaeric) [inline app/models/service_update.rb:23]: As `updates` should always be an array, mapping over it should be fine. No need to early return.

File: config/routes/placements.rb [hunk 18]
```diff
@@ -1,3 +1,7 @@
-scope module: :placements, as: :placements, constraints: { host: ENV["PLACEMENTS_HOST"] } do
+scope module: :placements,
+      as: :placements,
+      constraints: {
+        host: ENV["PLACEMENTS_HOST"]
+      } do
```
Comment (Nitemaeric) [inline config/routes/placements.rb:5]: We should avoid adding any changes that aren't relevant to this specific PR objective.

File: app/models/service_update.rb (hunk 17 above)
Comment (elceebee) [inline app/models/service_update.rb:23]: You are coding very defensively, which is a very good thing when you are consuming something changeable or depending on asynchronous jobs. But we don't want to code defensively against things we are responsible for (ie, our file structure). Our defence against someone deleting a file should be the tests. In this situation, if someone deletes the file, the tests would pass (because of the early returns), but the feature would be broken. 

And a warning, I'm pedantic about naming -- I don't think these are `params`, they are individual updates. And the `yaml_file` function returns a path, not a `yaml_file`

There are lots of ways to write this, but one alternative would be:

```suggestion
    path = file_path(service:)
    updates = YAML.load_file(path) || []

    updates.map { |service_update| new(service_update) }
```

File: app/models/service_update.rb (hunk 10 above)
Comment (elceebee) [reply app/models/service_update.rb:29]: Agreed.

File: spec/models/service_update_spec.rb (hunk 4 above)
Comment (elceebee) [reply spec/models/service_update_spec.rb:57]: I think this is fine, it's not actually reading the file, just checking that the function returns the the expected value.

File: spec/system/home_page_spec.rb [hunk 19]
```diff
@@ -1,9 +1,7 @@
 require "rails_helper"
 
 RSpec.feature "Home Page" do
-  after do
-    Capybara.app_host = nil
-  end
+  after { Capybara.app_host = nil }
```
Comment (elceebee) [inline spec/system/home_page_spec.rb:4]: Did the linter do this? If so, fine (odd it wasn't picked up before). 

If it didn't, I don't think we should introduce changes to unrelated files in a PR. Both syntaxes are valid and the change doesn't have anything to do with the task, so there is no need to change it.

File: config/routes.rb [hunk 20]
```diff
@@ -14,6 +14,11 @@
   get("/auth/developer/sign-out", to: "sessions#signout")
   post("/auth/developer/callback", to: "sessions#callback")
 
+  resources :service_updates,
+            only: %i[index],
+            path: "service-updates",
```
Comment (elceebee) [inline config/routes.rb:19]: Is this the pattern we will want to use for all paths (ie, hyphen instead of snake case?). If so, we should change it at the top of the app in an initialiser (`ActionDispatch`) rather than for each individual route.

**Review by elceebee — CHANGES_REQUESTED** [review]
Nice work, just a few comments. If anything is unclear, let me know, happy to discuss.

File: Gemfile.lock [hunk 21]
```diff
@@ -499,8 +500,7 @@ GEM
     zeitwerk (2.6.12)
 
 PLATFORMS
-  arm64-darwin-22
-  arm64-darwin-23
+  x86_64-darwin-22
```
Comment (elceebee) [inline Gemfile.lock:503]: Please don't delete `arm64-darwin-23` -- that's my M3.

Comment (elceebee) [reply Gemfile.lock:503]: How odd! We could get into a weird back and forth where every time I run bundle, it adds it back in, every time you run bundle it removes it :)

File: app/assets/components/service_update/view.rb (hunk 9 above)
Comment (ollietreend) [inline app/assets/components/service_update/view.rb:35]: I think a lot of the heavy lifting can be done using this DfE gem:
https://github.com/DFE-Digital/govuk-markdown

It uses Redcarpet configured to produce HTML with the Design System class names.

For example:

```ruby
GovukMarkdown.render("[A link](/foo)")
```

Will output:

```html
<p class="govuk-body-m">
  <a href="/foo" class="govuk-link">A link</a>
</p>
```

It seems like we could avoid a lot of the manual HTML tag config done within this ViewComponent if we used that instead.

**Review by ollietreend — CHANGES_REQUESTED** [review]

File: app/helpers/application_helper.rb [hunk 22]
```diff
@@ -1,5 +1,5 @@
 module ApplicationHelper
-  def current_service
+  def current_service(request)
```
[the remark this answers is not in the document]
Comment (elceebee) [reply app/helpers/application_helper.rb:2]: The controller has access to the request, so I think you must mean that you need to pass it when calling `current_service` in the view file? There are other ways to access this in views -- You could set a constant like `@service_name ||= current_service.to_s` in the controller and passing that to the view. Or add a view helper that takes the request. 

Making the change like this means we will have to unnecessarily pass the request every time we use `current_service` in a controller -- which will probably be pretty frequently.

**Review by ollietreend — APPROVED** [review]

**Review by elceebee — APPROVED** [review]

File: db/claims_service_updates.yml (hunk 3 above)
Comment (SocFoot) [reply db/claims_service_updates.yml:3]: Is `price:1.5` a float or a decimal? That's why you need the type.

**Review by SocFoot — APPROVED** [review]
